### PR TITLE
Avoid conversion of Arrays to streams when looping

### PR DESCRIPTION
**Pull request by DavideD**
@Sanne are these changes going to be helpful?

### CODE REVIEW

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java [hunk 1]
```diff
@@ -501,16 +500,15 @@ else if ( isAllOrDirtyOptimisticLocking() && loadedState != null ) {
...
-				)
+		return loop( 0, span,
+					 table -> deleteReactive(
+							 id,
+							 version,
+							 span - table - 1,
+							 deleteStrings[span - table - 1],
+							 session,
+							 state
+					 )
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:511]: But doesn't this reverse the order of the iteration?

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java [hunk 2]
```diff
@@ -728,19 +726,20 @@ else if ( !delegate().isModifiableEntity( entry ) ) {
...
+								fields,
+								oldFields,
+								table == 0 ? rowId : null,
+								propsToUpdate,
+								table,
+								oldVersion,
+								updateStrings[table],
+								session
+						)
+						: voidFuture()
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:742]: I'm not convinced by this one. I'm not sure what implication passing all those extra `voidFuture()`s has. Perhaps it's fine, but I would need to dig a lot deeper into the code.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java [hunk 3]
```diff
@@ -143,8 +142,12 @@ public static void logSqlException(Throwable t, Supplier<String> message, String
...
 	 * </pre>
 	 */
-	public static <T> CompletionStage<Void> loop(T[] array, Function<T,CompletionStage<?>> consumer) {
-		return loop( Arrays.stream(array), consumer );
+	public static <T> CompletionStage<Void> loop(T[] array, Function<T, CompletionStage<?>> consumer) {
+		AtomicInteger index = new AtomicInteger();
+		return AsyncTrampoline.asyncWhile( () -> {
+			int i = index.getAndIncrement();
+			return consumer.apply( array[i] ).thenApply( v -> i < array.length - 1 );
+		} );
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java:150]: Are we really certain that creating an `AtomicInteger` is less expensive than creating a stream? I mean, it certainly sounds _possible_, but it's not _obviously_ true.

**Review by gavinking — COMMENTED** [review]
I mean, I'm not convinced on first sight, but I guess if you really dug into this??

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java (hunk 3 above)
Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java:156]: Remember: "creating a stream" in this case just means making a little wrapper for an array.

**Comment by DavideD** [comment]
We can wait about merging this one and do proper performance testing.
There are other things that I want to try. 

I'm going to close the pr for now

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java (hunk 1 above)
Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:511]: That was what the original code was doing

Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/persister/entity/impl/ReactiveAbstractEntityPersister.java:511]: Oh, apologies, I missed that you had changed it to use `span - table - 1`. That's better, actually.

**Pull request closed without merging**